myia/parser.py

```diff
@@ -72,34 +105,30 @@ class Environment:
     convert it into a Myia object e.g. create a Myia constant or parse a Python
     function into a Myia function.
 
+    Note that the parser uses the mapping for ``operator.<op>`` to determine
+    which primitive to use for operation ``<op>``.
+
     The environment is also in charge of resolving variable names that could
     not be resolved in the function's local and global namespace i.e. for
     built-ins and for undefined variable names.
 
-    Lastly, the environment has a mapping from AST nodes to Myia nodes that the
-    parser uses e.g. for mapping binary operators to the correct primitives.
-
     Functions parsed in different environments will have no nodes in common.
     Functions parsed in the same environment will use the same Myia objects for
     the same Python objects.
 
     """
 
-    def __init__(self, object_map: Dict[int, ANFNode],
```

Comment:
I remember that the problem with `ast_map` was that that there are combinations of AST nodes that map to a single operator (e.g. assign to a tuple means unpacking). That said, I do think there is place for two mappings: One of them maps Python syntax to nodes, the other one maps Python objects to nodes. One is immutable during parsing, the other one is not. Folding them together makes the code harder to understand IMO.

Reply:
They are not folded together. I made a static `ast_map` at the top of the file that maps AST node types to `operator.<op>`. To resolve `ast.Add`, the parser first looks up what operator it is in `ast_map` then it looks up the operator in `object_map`. I think that's sensible, since you're probably never going to want `a + b` to differ from `operator.add(a, b)`.